### PR TITLE
refactor: abstract common functions in `Keys` and `KeysRange`

### LINKED ISSUE #987: refactor: abstract common functions in `Keys` and `KeysRange`
New description:
Instead of replacing `Keys` with `KeysRange` the common functions in them will be abstracted out.

Old description: 
[`Keys`](https://github.com/near/near-sdk-rs/blob/2400a34272a9ecfe6f020c16fda6cdf2029543cf/near-sdk/src/store/tree_map/iter.rs#L296) is a sub case of [`KeysRange`](https://github.com/near/near-sdk-rs/blob/2400a34272a9ecfe6f020c16fda6cdf2029543cf/near-sdk/src/store/tree_map/iter.rs#L386) and can be replaced with it. 

Replacing `Keys` with `KeysRange` will remove duplicate code. Please see https://github.com/near/near-sdk-rs/pull/964#discussion_r1032590540 for more details.

### PR DESCRIPTION
**Pull request by askoa**
Closes #987

### CODE REVIEW

File: near-sdk/src/store/tree_map/iter.rs [hunk 1]
```diff
@@ -584,7 +398,11 @@ where
 
     fn size_hint(&self) -> (usize, Option<usize>) {
         let len = self.length as usize;
-        (0, Some(len))
+        (len, Some(len))
```
Comment (austinabell) [inline near-sdk/src/store/tree_map/iter.rs:401]: So the main concern with this is that this is stating that there is an exact amount of elements in this iterator, which is not the case for `KeysRange`. I do think that this difference can be useful to users, if they want to use `Keys` as an exact size iterator.

I would recommend trying to keep `Keys` and `KeysRange` as separate types, but just share the internal logic with shared types or abstracting the methods to just take the raw components.

File: near-sdk/src/store/tree_map/mod.rs [hunk 2]
```diff
@@ -9,10 +9,10 @@ use crate::store::LookupMap;
 use crate::{env, IntoStorageKey};
 use borsh::{BorshDeserialize, BorshSerialize};
 pub use entry::Entry;
-pub use iter::{Iter, IterMut, Keys, Range, RangeMut, Values, ValuesMut};
+pub use iter::{Iter, IterMut, KeysRange, Range, RangeMut, Values, ValuesMut};
```
Comment (austinabell) [inline near-sdk/src/store/tree_map/mod.rs:12]: Oh... interesting... seems like the `KeysRange` type was not exported previously. That was a mistake :D

Comment (askoa) [reply near-sdk/src/store/tree_map/mod.rs:12]: You want both `Keys` and `KeysRange` in here?

File: near-sdk/src/store/tree_map/iter.rs (hunk 1 above)
Comment (askoa) [reply near-sdk/src/store/tree_map/iter.rs:401]: Cool! I'll add `Keys` back and have shared methods.

File: near-sdk/src/store/tree_map/mod.rs (hunk 2 above)
Comment (austinabell) [reply near-sdk/src/store/tree_map/mod.rs:12]: oh, actually. Sorry for the false positive here. No, KeysRange is only used internally; it was correct not to be exposed.

File: near-sdk/src/store/tree_map/iter.rs [hunk 3]
```diff
@@ -294,6 +294,116 @@ where
     }
 }
 
+fn find_min<'a, K>(
+    tree: &'a Tree<K>,
+    root: Option<&FreeListIndex>,
```
Comment (austinabell) [inline near-sdk/src/store/tree_map/iter.rs:299]: small nit: the root is coming from the tree, so there should be no need to pass in the root in this and `find_max`, right?

Comment (askoa) [reply near-sdk/src/store/tree_map/iter.rs:299]: I think its needed. Its the root of the `find_min` and not root of the tree. See below.

https://github.com/near/near-sdk-rs/blob/1351cad09653a8a333be7509fb9941fcbf79bc61/near-sdk/src/store/tree_map/iter.rs#L366

Comment (austinabell) [reply near-sdk/src/store/tree_map/iter.rs:299]: oops! I missed that when I skimmed this

**Review by austinabell — APPROVED** [review]
LGTM! Thanks :) 

@itegulov do you want to have a skim through to make sure I didn't miss something nuanced here?

**Review by itegulov — APPROVED** [review]

**Pull request merged**